docs/RFCS/gateway_batch.md

```diff
@@ -0,0 +1,147 @@
+ground work for major server-side performance improvements.
+
+# Detailed design
+
+The proposition below assumes #1988.
+
+The summary might make this seem like a straightforward change, but there are
+some subtleties involved.
+
+Naively, `DistSender` would be in charge of breaking a `Batch` up into pieces
+transparently. This is fine for non-transactional batches, but consider a
+`Transaction`: the happy case is that everything affects only one `Range` and
+the `Batch` can be sent to that `Range` as a whole. If that is not the case and
+we end up with several chunks of `Batch`, it's clear that the one containing
+`EndTransaction` must be sent last. This is nothing the `DistSender` should
+worry about, and so from that perspective, splitting the `Batch` is
+`TxnCoordSender`'s job.
+
+But it's desirable to keep the actual "cutting" of the batch in `DistSender`
+since that's where the range descriptors are cached and looked up. It's also
+desirable to have the cutting and sending in one context, so that all the
+information is available to deal with errors (and specifically cache
+invalidation) appropriately.
+
+I think the snippet interface is a good starting point for discussion:
+```go
+// SendBatch is called by (*TxnCoordSender).Send().
+func (ds *DistSender) SendBatch(b *proto.BatchRequest, preHook func([]*proto.BatchRequest)) proto.BatchResponse {
+  // Retry loop here.
+  // ...
+
+    // Chop up the batch. This is already in a retry loop.
+    var chopped []*proto.BatchRequest = ds.Split(b)
+    // Execute the callback, which may rearrange some items and update
+    // `EndTransaction`.
+    // preHook **must** be idempotent! It may be executed multiple times if
+    // stale descriptors causes a retryable error.
+    b = preHook(b)
+
+    // Send the reordered batch requests one by one, much like ds.Send()
+    // ...
+
+  // Recombine the multi-range requests back into one.
+  // ...
+
+  return response
+}
+```
+
+That would cover the above bases: `TxnCoordSender` gets to inspect the chunks of
+the batch and reorders them such that `EndTransaction` comes last; it also makes
+sure that `EndTransaction` has the complete list of intents. Another reasonable
+optimization, namely sending chunks of the batch in parallel, could be carried
+out similarly: In this case, `preHook` would not only operate on the slice, but
+return a `Plan` consisting of different stages, each containing multiple chunks
+(extremes: each stage consisting of one one-element batch = sequential querying
+as we do it now; one stage of `N` batches (to different ranges) = completely
+parallel). Non-transactions would go completely parallel, transactions are a
+bit more involved (going parallel might lay down a lot of intents
+and do a lot of work for nothing) but will always have a last stage with the
+chunk containing `EndTransaction`.
+
+## Placeholders and Multi-Range requests
+
+We need to introduce a placeholder entry in a `Batch` (`nil`?) so that the
```

Comment:
This paragraph is unclear. I would remove this paragraph, and then after the discussion of `Combinable` results, add a new paragraph something like:

Therefore, when we split a batch, each new batch has the same size as the original batch, padded with placeholder values (`nil`?) as necessary. Each request has the same index in the new batch as it had in the original batch. Most indexes will have a non-nil value in only one batch, but a request that spans multiple ranges will be duplicated into more than one batch.